testing/testing.gni

```diff
@@ -6,153 +6,258 @@ import("//build/compiled_action.gni")
-      "--isolate_snapshot_instructions=" +
-          rebase_path("$fixtures_location/isolate_snapshot_instr"),
-      rebase_path("$fixtures_location/$kernel_out"),
+      "--vm_snapshot_data=" + rebase_path(vm_snapshot_data),
+      "--vm_snapshot_instructions=" + rebase_path(vm_snapshot_instr),
+      "--isolate_snapshot_data=" + rebase_path(isolate_snapshot_data),
+      "--isolate_snapshot_instructions=" + rebase_path(isolate_snapshot_instr),
+      rebase_path(invoker.dart_kernel),
     ]
 
-    deps = [
-      ":$fixtures_kernel_target_name",
-    ]
+    forward_variables_from(invoker, [ "deps" ])
+  }
+}
+
+# Generates a kernel or AOT snapshot as necessary from the main Dart file.
+# Other Dart dependencies referenced by that main Dart file will be tracked.
+#
+# Arguments:
+#
+#     dart_main (required): The path to the main Dart file.
+template("dart_snapshot") {
+  assert(defined(invoker.dart_main), "The main Dart file must be specified.")
+
+  testonly = true
+
+  dart_snapshot_kernel_target_name = "dart_snapshot_kernel_$target_name"
+  dart_snapshot_kernel_path = "$target_gen_dir/assets/kernel_blob.bin"
+  dart_snapshot_kernel(dart_snapshot_kernel_target_name) {
+    dart_main = invoker.dart_main
+    dart_kernel = dart_snapshot_kernel_path
+  }
+
+  snapshot_deps = [ ":$dart_snapshot_kernel_target_name" ]
+
+  if (is_aot_test) {
+    dart_snapshot_aot_target_name = "dart_snapshot_aot_$target_name"
+    dart_snapshot_aot(dart_snapshot_aot_target_name) {
+      dart_kernel = dart_snapshot_kernel_path
+      deps = [
+        ":$dart_snapshot_kernel_target_name",
+      ]
+    }
+    snapshot_deps += [ ":$dart_snapshot_aot_target_name" ]
   }
 
   group(target_name) {
     testonly = true
-    deps = [
-      ":$fixtures_source_set_target_name",
-    ]
-    if (flutter_runtime_mode == "profile" || flutter_runtime_mode == "release") {
-      deps += [ ":$fixtures_aot_target_name" ]
-    } else {
-      deps += [ ":$fixtures_kernel_target_name" ]
+
+    deps = snapshot_deps
+  }
+}
+
+# Copies a (potentially empty) list of fixtures to the fixtures directory for
+# the unit test.
+#
+# Arguments:
+#
+#     fixtures (required): The list of fixtures to copy.
+template("copy_fixtures") {
+  testonly = true
+
+  assert(defined(invoker.fixtures), "The test fixtures must be specified.")
+
+  has_fixtures = false
+  foreach(fixture, invoker.fixtures) {
+    has_fixtures = true
+  }
+
+  if (has_fixtures) {
+    copy(target_name) {
+      sources = invoker.fixtures
+      outputs = [
+        "$target_gen_dir/assets/{{source_file_part}}",
+      ]
+    }
+  } else {
+    group(target_name) {
+      # The copy target cannot accept an empty list.
+    }
+  }
+}
+
+# Specifies the fixtures to copy to a location known by the specific unit test.
+# Test executable can only depend on one such target.
+#
+# Arguments:
+#
+#     fixtures (required): The list of test fixtures. An empty list may be
```

Comment:
I wonder why we must provide an empty list `[]`? For example, is
```
test_fixtures("runtime_fixtures") {
  dart_main = "fixtures/runtime_test.dart"
   fixtures = []
}
```
a little unnecessarily verbose and confusing than
```
test_fixtures("runtime_fixtures") {
  dart_main = "fixtures/runtime_test.dart"
}
```
?

If there's a nice reason to put an empty list here, maybe it's worth to write down that reason in this documentation.

Reply:
The reason for the empty list is a GN quirk where if the template expansion does not need any arguments (as it is in this case where the `target_gen_dir` used to compute the fixtures directory), the `invoker` scope goes unused in the expansion and GN warns of this and errors out.

Reply:
You can have either one or both the arguments in the template and it will work fine. If you have none, then you'll see the unused invoker scope error.

Reply:
I see, thanks! It would be nice to add `You can have either one or both the arguments in the template and it will work fine. If you have none, then you'll see the unused invoker scope error.` to the comment so in a few months for now, I can still figure this out by reading the comments instead of `git blame` :smile: